File: validation/check_search.js
'use strict';
const assert = require('node:assert/strict');
const {readFileSync} = require('node:fs');
const {join} = require('node:path');
const {normalize, prepare, search, snippet} = require('../assets/search.js');
const data = JSON.parse(readFileSync(join(__dirname,'../docs/search-index.json'),'utf8'));
const records = prepare(data.records);
const urls = query => search(records,query).map(hit=>hit.record.url);
assert.ok(urls('CMP29').length > 0);
for (const query of ['MP29','MP 29','cmp-29','MP–29','CMP−29']) {
  assert.deepEqual(urls(query), urls('CMP29'), query);
}
assert.ok(urls('CMP29 2027').some(url=>url.startsWith('tac-advice-2027-2028.html#')));
assert.ok(urls('CPUE').some(url=>url.startsWith('assessment/SC14.html#')));
assert.ok(urls('recruit').length > 0);
assert.equal(urls('zzzxxyynonexistent').length, 0);
assert.equal(urls('x').length, 0);
assert.equal(urls('!@#$%').length, 0);
assert.equal(normalize('Péru'), 'peru');
const text = 'Unrelated introduction. '.repeat(30) + 'MP29 applies the candidate rule to the index.';
assert.match(snippet(text,'CMP29'), /MP29 applies/);
assert.equal(snippet(text,'CMP29'),snippet(text,'MP 29'));
const fixture = prepare([{title:'Survey',heading:'Index',url:'survey.html',text:'Peru CPUE observations'}]);
assert.equal(search(fixture,'Peru acoustic').length,0);
assert.equal(search(fixture,'Peru CPUE').length,1);
assert.equal(new Set(urls('TAC')).size, urls('TAC').length);
console.log('PASS: CMP aliases, excerpts, TAC/assessment retrieval, prefix and multiple-term matching, empty results and deduplication');

File: validation/check_search.py
"""Check search coverage, destinations and navigation against the built site."""
from html.parser import HTMLParser
from pathlib import Path
from urllib.parse import unquote, urlsplit
import json
import sys

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))
from build_search import SearchText


class IDs(HTMLParser):
    def __init__(self):
        super().__init__()
        self.ids = set()

    def handle_starttag(self, tag, attrs):
        attrs = dict(attrs)
        if 'id' in attrs:
            self.ids.add(attrs['id'])


index = json.loads((ROOT/'docs/search-index.json').read_text())
documents = {d['url'] for d in index['documents']}
assert 'tac-advice-2027-2028.html' in documents
assert 'assessment/SC14.html' in documents
assert 'assessment/loo.html' in documents
assert 'search.html' not in documents
assert not any('annex/' in d or 'kobe-100-trajectories' in d for d in documents)
assert len(documents) == len(index['documents'])
ids = {}
for document in documents:
    parser = IDs()
    parser.feed((ROOT/'docs'/document).read_text())
    ids[document] = parser.ids
for record in index['records']:
    destination = urlsplit(record['url'])
    assert destination.path in documents, record['url']
    assert not destination.fragment or unquote(destination.fragment) in ids[destination.path], record['url']
    assert destination.fragment != 'quarto-content', record['url']
    assert record['text'].strip()
    assert 'Search wiki' not in record['text']
for source in (ROOT/'content').glob('*.md'):
    rendered = (ROOT/'docs'/f'{source.stem}.html').read_text()
    assert 'class="wiki-search" action="search.html"' in rendered, source
    assert '<script defer src="search.js"></script>' in rendered, source
for asset in ('search.js', 'search.css'):
    assert (ROOT/'assets'/asset).read_bytes() == (ROOT/'docs'/asset).read_bytes(), asset
fixture = SearchText()
fixture.feed('<title>Example</title><nav>Exclude nav</nav><main><h1 id="top">Fish</h1><p>Visible text</p><script>Exclude script</script><section id="survey"><h2>Survey</h2><p>CPUE</p></section><footer>Exclude footer</footer></main>')
assert [s['anchor'] for s in fixture.sections] == ['top', 'survey']
assert all('Exclude' not in s['text'] for s in fixture.sections)
print(f"PASS: {len(documents)} documents, {len(index['records'])} passages, all destinations and wiki search forms checked")
